packages/theme/src/cli/utilities/theme-environment/local-assets.ts

```diff
@@ -94,35 +94,143 @@ function isCompiledAssetRequest(event: H3Event) {
-        .sort(([key1], [key2]) => key1.localeCompare(key2))
-    }
+const compiledStylesheets = new Map<string, {checksum: string; content: string}>()
+const compiledJavaScripts = new Map<string, {checksum: string; content: string}>()
+
+function handleStylesCss(ctx: DevServerContext, event: H3Event) {
+  const sectionFiles = getLiquidFilesByKind(ctx, 'section')
+  const blockFiles = getLiquidFilesByKind(ctx, 'block')
+  const snippetFiles = getLiquidFilesByKind(ctx, 'snippet')
+
+  const stylesheets = ['/*** GENERATED LOCALLY ***/\n']
 
-    const sectionFiles = getLiquidFilesFromDir('sections')
-    const blockFiles = getLiquidFilesFromDir('blocks')
-    const snippetFiles = getLiquidFilesFromDir('snippets')
+  for (const [, file] of [...sectionFiles, ...blockFiles, ...snippetFiles]) {
+    const stylesheet = getCompiledStylesheet(file)
+    if (stylesheet) stylesheets.push(stylesheet)
+  }
+
+  const stylesheet = stylesheets.join('\n')
+
+  return serveStatic(event, {
+    getContents: () => stylesheet,
+    getMeta: () => ({type: 'text/css', size: stylesheet.length, mtime: new Date()}),
+  })
+}
 
-    const stylesheets = ['/*** GENERATED LOCALLY ***/\n']
-    for (const [, file] of [...sectionFiles, ...blockFiles, ...snippetFiles]) {
-      const stylesheet = getCompiledStylesheet(file)
-      if (stylesheet) stylesheets.push(stylesheet)
+function handleBlockScriptsJs(ctx: DevServerContext, event: H3Event, kind: 'block' | 'snippet' | 'section') {
+  const blockFiles = getLiquidFilesByKind(ctx, kind)
+  const kinds = `${kind}s`
+
+  const javascripts = [
+    `
+      /*** GENERATED LOCALLY ***/
+
+      (function () {
+        var __${kinds}__ = {};
+
+        (function () {
+          var element = document.getElementById("${kinds}-script");
+          var attribute = element ? element.getAttribute("data-${kinds}") : "";
+          var ${kinds} = attribute.split(",").filter(Boolean);
+
+          for (var i = 0; i < ${kinds}.length; i++) {
+            __${kinds}__[${kinds}[i]] = true;
+          }
+        })();
+    `,
+  ]
+
+  for (const [key, file] of blockFiles) {
+    const baseName = key.split('/').pop()?.replace('.liquid', '')
+    const javascript = getCompiledJavaScript(file)
+
+    if (baseName && javascript) {
+      /**
+       * For sections, this approach isn't completely accurate, as we'd have
+       * something like this:
+       * ```
+       *    if (!__sections__['header'] && !window.DesignMode) return;
+       * ```
+       * However, this works well enough for most cases. It's preferable to have
+       * a consistent JavaScript workflow across sections, blocks, and snippets,
+       * especially when considering hot reloading. This only affects JavaScript
+       * sections in the theme editor and doesn't impact the main runtime.
+       */
+      javascripts.push(`
+        (function () {
+          if (!__${kinds}__["${baseName}"] && !Shopify.designMode) return;
```

Comment:
So this doesn't run in OSE?

Reply:
Actually, it always runs on OSE (I also don't like this condition so much, but I'm trying to be as accurate as possible)